### PR TITLE
Adding testing framework documentation 

### PR DESCRIPTION
**Pull request by axiomcura**
_modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_

# Description

Thank you for your contribution to CytoSnake!
Please _succinctly_ summarize your proposed change.
What motivated you to make this change?

Please also link to any relevant issues that your code is associated with.

## What is the nature of your change?

This PR adds `CytoSnake` testing framework documentation. A preview of this PR  can be viewed here: https://cytosnake-testing-docs.readthedocs.io/en/latest/

Other changes:
- added logo to the website (top-left corner)

- [ ] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [X] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [X] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [X] My code follows the style guidelines of this project.
- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -30,4 +30,4 @@
 # https://www.sphinx-doc.org/en/master/usage/configuration.html#options-for-html-output
 
 html_theme = "furo"
-html_static_path = ["_static"]
+html_logo = "../logo/just-icon.svg"
```
Comment (gwaybio) [inline docs/conf.py:33]:
```suggestion
html_logo = "../logo/just-icon.svg"

```

File: docs/requirements.txt [hunk 2]
```diff
@@ -1,7 +1,6 @@
-sphinx
+# sphinx
```
Comment (gwaybio) [inline docs/requirements.txt:1]: just checking if this should be commented

Comment (gwaybio) [reply docs/requirements.txt:1]: Also, I don't see where you've specified the furo version

File: tests/README-tests.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+## CytoSnake Testing General Documentation
+
+<p align="center">
+  <img src="../docs/images/testing_diagram.svg", alt="image showing CytoSnake's Testing Suite">
+</p>
+
+>This diagram illustrates the testing framework of `CytoSnake`, showcasing the main objectives of each core testing module.
+
+Welcome to the testing documentation for `CytosSnake`, where you'll learn the essential principles of crafting effective and insightful tests.
```
Comment (gwaybio) [inline tests/README-tests.md:9]:
```suggestion
Welcome to the testing documentation for `CytoSnake`, where you'll learn the essential principles of crafting effective and insightful tests.
```

Also, this sentence implies that the audience for this documentation is developers. Is this true?

If so, it might be worth adding a sentence or two specifying who should read this document and why.

File: tests/README-tests.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Our tests are centered around ensuring that CytoSnake stands as a robust and dependable software tool, capable of generating reproducible workflows for single-cell morphology profiles.
```
Comment (gwaybio) [inline tests/README-tests.md:10]:
```suggestion
Our tests are centered around ensuring that `CytoSnake` stands as a robust and dependable software tool, capable of generating reproducible workflows for single-cell morphology profiles.
```

To use consistent backticks

File: tests/README-tests.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+In this documentation, you will gain insight into creating tests within CytoSnake, ensuring that every user who uses the `CytoSnake` can get a reliable experience.
```
Comment (gwaybio) [inline tests/README-tests.md:12]:
```suggestion
In this documentation, you will gain insight into creating tests within `CytoSnake`, ensuring that every user who uses your pipelines within `CytoSnake` can have a reliable experience.
```

True?

File: tests/README-tests.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+- [Unit testing](../docs/unit-tests.md)
+- [Workflow testing](../docs/workflow-tests.md)
+
+## Rationale and Motivation
+
+Below are the rationals and motivations behind generating tests for `CytoSnake`
+
+### Performance
+
+Remarkable technological advances have occurred within the field of high-throughput microscopy imaging, generating a surmount of single-cell morphology readouts, also known as single-cell morphology profiles.
```
Comment (gwaybio) [inline tests/README-tests.md:26]:
```suggestion
Remarkable technological advances have occurred within the field of high-throughput microscopy imaging, generating a lot of single-cell morphology readouts, also known as single-cell morphology profiles.
```

File: tests/README-tests.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+With this increase in data, there is a high need for high-performing software tools that can not only scale with large volumes of single-cell morphology profiles but also create a tool that can orchestrate various analytical tools and make them reproducible.
+Our tests in `CytoSnake` are designed to align with our goals of scale, performance, and orchestration. By rigorously evaluating core functionality, we ensure that users can quickly obtain reproducible conclusions from their data, enabling faster insights and analysis.
```
Comment (gwaybio) [inline tests/README-tests.md:28]:
```suggestion
Our tests in `CytoSnake` are designed to align with our goals of scale, performance, and orchestration.
By rigorously evaluating core functionality, we ensure that users can quickly obtain reproducible conclusions from their data, enabling faster insights and analysis.
```

File: tests/README-tests.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+### Reproducibility
+
+Reproducibility is an integral part of Cytosnake ensuring the reliable robust conclusions.
```
Comment (gwaybio) [inline tests/README-tests.md:32]:
```suggestion
Reproducibility is an integral part of `CytoSnake` ensuring the reliable robust conclusions.
```

Consistent backticks and capitalization

File: tests/README-tests.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+Our tests are designed to ensure reproducibility by monitoring `CytoSnake`'s functionality across different scenarios.
+This means that CytoSnake will be subjected to various inputs and conditions to test actively identifying and capturing any inconsistencies with the generated outputs.
```
Comment (gwaybio) [inline tests/README-tests.md:34]:
```suggestion
This means that `CytoSnake` will be subjected to various inputs and conditions to test actively identifying and capturing any inconsistencies with the generated outputs.
```

Please check all is consistent (i'm going to stop adding suggestions :) )

File: tests/README-tests.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+Our testing design aims to offer researchers in the community a sense of reliability, ensuring they can consistently achieve reproducible results that reinforce robust scientific conclusions.
+Overall, `CytoSnake`'s tests are designed to not only reach robust conclusions but also accelerate progress in the field of Cell biology, enabling faster and more reliable advancements.
```
Comment (gwaybio) [inline tests/README-tests.md:36]:
```suggestion
Overall, `CytoSnake`'s tests are designed to not only reach robust conclusions but also accelerate progress in the field of Quantitative Cell Biology, enabling faster and more reliable advancements.
```

File: tests/README-tests.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+### User-friendliness and Portability
+
+While prioritizing reproducibility is crucial, we also intend to make `CytoSnake` accessible to a broader audience, ensuring its user-friendliness and portability.
+Our tests are deployed in GitHub Actions, allowing for multi-platform testing across different operating systems, ensuring `CytoSnake` is system agnostic.
+In addition, we use the `Poetry` package to manage dependencies, assuring that the software can be easily installed.
+We leverage the `conda` environment manager to encapsulate dependencies within `CytoSnake`'s modules allowing it to be portable and consistent across all systems.
+Ultimately, our tests aim to provide users with an easy and enjoyable experience while using `CytoSnake`, by providing the convenience of being portable and requiring little effort to use it.
```
Comment (gwaybio) [inline tests/README-tests.md:44]:
```suggestion
Ultimately, our tests aim to ensure that users have an easy and enjoyable experience while using `CytoSnake`, by providing the convenience of being portable and requiring little effort to use it.
```

The tests won't provide the experience, just ensure it - right?

File: tests/README-tests.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+### Tech stack
+
+Our testing suite uses [`pytest`](https://docs.pytest.org/en/7.4.x/) a very powerful, flexible and easy-to-use testing framework.
+Its user-friendliness enables developers to write and execute tests easily, providing rapid development.
+Additionally, Pytest's modular fixtures and extensive customization provide great opportunities with CytoSnake's development workflow, enhancing code quality and adaptability while maintaining a smooth and efficient testing process.
+
+We also use `Code Coverage` as a metric for our tests to ensure high code quality and prevent code decay.
+Through development, we use code coverage as a monitoring tool that ensures that `CytoSnake`'s codebase remains robust against decay, preventing any bugs that might arise over time.
+This approach ensures that all functionalities of `CytoSnake` are thoroughly tested, making it a dependable software tool. Additionally, developers have the flexibility to create their own datasets, as long as they adhere to CytoSnake's established standards.
```
Comment (gwaybio) [inline tests/README-tests.md:61]:
```suggestion
This approach ensures that all functionalities of `CytoSnake` are thoroughly tested, making it a dependable software tool.
Additionally, developers have the flexibility to create their own datasets, as long as they adhere to CytoSnake's established standards.
```

File: tests/README-tests.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+### Dataset
+
+We use structured datasets that attempt to emulate different uses of CytoSnake. These datasets have different versions to cover various situations that might occur when users interact with CytoSnake.
```
Comment (gwaybio) [inline tests/README-tests.md:65]:
```suggestion
We use structured datasets that attempt to emulate different uses of CytoSnake. 
These datasets have different versions to cover various situations that might occur when users interact with CytoSnake.
```

Comment (gwaybio) [reply tests/README-tests.md:65]: Also, what is a "structured" dataset? Might want to clarify

File: tests/README-tests.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+### Clear and Descriptive test
+
+In `CytoSnake`, well-structured testing documentation plays an important role.
+It simplifies the software experience for both users and developers by providing a clear understanding of what is being tested within `CytoSnake`.
+We believe that providing easy-to-follow documentation generated from users can serve as a valuable resource. Documentation not only helps in understanding the testing processes but also offers insights into the overall functionality of `CytoSnake`.
```
Comment (gwaybio) [inline tests/README-tests.md:71]:
```suggestion
We believe that providing easy-to-follow documentation generated from users can serve as a valuable resource.
Documentation not only helps in understanding the testing processes but also offers insights into the overall functionality of `CytoSnake`.
```

File: tests/README-tests.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+By capturing diverse scenarios and potential edge cases, this documentation becomes a reliable reference for users seeking to understand `CytoSnake`'s capabilities.
```
Comment (gwaybio) [inline tests/README-tests.md:72]:
```suggestion
By capturing diverse scenarios and potential edge cases, our testing suite also becomes a reliable reference for users seeking to understand `CytoSnake`'s capabilities.
```

File: tests/functional/README-functional.md [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+Functional tests are essential for evaluating `CytoSnake`'s behavior in accurately executing its intended functions with specific input parameters.
+They primarily aim to validate the functionality of the program, ensuring it meets the specified requirements.
+These tests involve feeding `CytoSnake`'s Command Line Interface (CLI) with various parameters to simulate user interactions and thoroughly monitor the user experience
+By conducting functional tests, developers can ensure that all user-inputted parameters function correctly within `CytoSnake`, leading to a more reliable and user-friendly application.
+Ultimately, the focus of these tests is to guarantee easy and simple user experience with `CytoSnake`.
+
+## Testing Environment
+
+Below is an image on how the testing environment:
```
Comment (gwaybio) [inline tests/functional/README-functional.md:30]:
```suggestion
```

I think this is redundant with the figure legend and title, but maybe there is key detail there that I'm missing.

File: tests/functional/README-functional.md [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+## Rational
+
+Functional tests are essential for evaluating `CytoSnake`'s behavior in accurately executing its intended functions with specific input parameters.
+They primarily aim to validate the functionality of the program, ensuring it meets the specified requirements.
+These tests involve feeding `CytoSnake`'s Command Line Interface (CLI) with various parameters to simulate user interactions and thoroughly monitor the user experience
+By conducting functional tests, developers can ensure that all user-inputted parameters function correctly within `CytoSnake`, leading to a more reliable and user-friendly application.
+Ultimately, the focus of these tests is to guarantee easy and simple user experience with `CytoSnake`.
+
+## Testing Environment
```
Comment (gwaybio) [inline tests/functional/README-functional.md:28]:
```suggestion
## Functional Testing Environment
```

File: tests/functional/README-functional.md [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+Below is an image on how the testing environment:
+
+<p align="center">
+  <img src="../../docs/images/functional-testing-diagram.png", alt="image showing CytoSnake's functional test env">
+</p>
+
+> **Functional Testing Environment**: Schematic dipicting `CytoSnake`'s functional testing envrionment.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:36]:
```suggestion
> **Functional Testing Environment**: Schematic depicting `CytoSnake`'s functional testing environment.
```

File: tests/functional/README-functional.md [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+
+Shown above is a basic illustration of the functional testing environment used by `CytoSnake`. This illustration introduces four fundamental components: the `CytoSnake` module(shown in blue-green), prepare_data() function (shown in light-green), functional test module (shown in yellow), and fixture (shown in red) working together.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:38]:
```suggestion
Shown above is a basic illustration of the functional testing environment used by `CytoSnake`.
This illustration introduces four fundamental components: the `CytoSnake` module(shown in blue-green), prepare_data() function (shown in light-green), functional test module (shown in yellow), and fixture (shown in red) working together.
```

File: tests/functional/README-functional.md [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+- **`CytoSnake module`**: is used to import `test_utils.py` module in order to use the `prepare_data()` function
```
Comment (gwaybio) [inline tests/functional/README-functional.md:40]: Maybe I'm not looking at the right place, but where in the diagram is the CytoSnake module? I'm looking for blue-green, but it would be helpful to also add a label if you're directly referencing it.

File: tests/functional/README-functional.md [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+- **`prepare_data()`**: allows developers to easily select data and transport them to the testing directory
```
Comment (gwaybio) [inline tests/functional/README-functional.md:41]: Is it also possible to add data to the test data directory? (Minor comment: Rather than being a circle, it should probably be a database symbol (3D cylinder))

File: tests/functional/README-functional.md [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+- **`functional test`** is where `CytoSnake` CLI is being tested.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:42]: Hmm, I'm not sure what you're referring to in the figure

File: tests/functional/README-functional.md [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+- **`fixture`**: Are [`pytest`](https://docs.pytest.org/en/6.2.x/fixture.html) fixtures that provide essential setup and teardown functionalities to make it easier for developers to form their tests.
+
+## Submitting Tests
+
+Adding functional tests is simple! You, as a developer, can create a new functional test within the `cytosnake/test/functional` directory and select a module where the test will be added.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:47]:
```suggestion
Adding functional tests is simple!
You, as a developer, can create a new functional test within the `cytosnake/test/functional` directory and select a module where the test will be added.
This is similar to adding tests in traditional software packages, but using a consistent pattern to help maintainability.
```

File: tests/functional/README-functional.md [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+Below, is an example to illustrate how to add a test in the `test_cli.py` module:
+
+```python
+# inside test_cli.py testing module
+@pytest.mark.positive # or can be negative as well
+def test_example(testing_dir) -> None:
+    """
+    [ Enter Documentation]
+    test type: (positive or negative)
+    rational:Small description why this test is being done.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:57]:
```suggestion
    rational: small description why this test is being done.
```

Adding space and making capitalization consistent

File: tests/functional/README-functional.md [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+    input documentation: what inputs are being used in your test.
+    """
+
+    # step1: prepare testing files and select input files
```
Comment (gwaybio) [inline tests/functional/README-functional.md:61]:
```suggestion
    # Step 1: prepare testing files and select input files
```

Consistent capitalization and spacing

File: tests/functional/README-functional.md [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+#### Preparation
+
+Preparing your dataset requires minimal effort with the help  `prepare_dataset` function.
+This function readily prepares the dataset using just two parameters: the name of the dataset and the target directory for the test execution.
+
+> **Note**: Thanks to the use of pytest's fixtures, the `testing_dir` variable already holds the path to the generated testing directory. As a result, users **are not required** to create their own testing directories when conducting functional tests.
+
+The dataset names are derived from the folder names within the `cytosnake/tests/datasets` directory.
+If the desired dataset is not available, developers can create their datasets; however, they must ensure they follow `CytoSnake's` input requirements.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:108]: Is there a link you can add to clarify what you mean by "input requirements"?

File: tests/functional/README-functional.md [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+
+After executing the `prepare_dataset` function, it generates an object called `DataFiles`, granting users full control over the selection of input files for `CytoSnake`.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:110]:
```suggestion
After executing the `prepare_dataset()` function, the fixture generates an object called `DataFiles`, granting users full control over the selection of input files for `CytoSnake`.
```

Please confirm that my suggestion is accurate.

File: tests/functional/README-functional.md [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+For instance, if the chosen dataset comprises multiple plates, users can select a specific plate using Python indexing, as demonstrated in the example above.
+Once the desired files are selected, users can proceed to prepare them for execution.
+
+#### Execution
+
+In the execution step, `CytoSnake` runs with the provided inputs within the testing directory.
+Fortunately, the `prepare_dataset` function not only generates the `DataFile` object but also informs the testing module about the execution location for `CytoSnake`.
```
Comment (gwaybio) [inline tests/functional/README-functional.md:117]:
```suggestion
Fortunately, the `prepare_dataset()` function not only generates the `DataFile` object but also informs the testing module about the execution location for `CytoSnake`.
```

I'm used to seeing functions referred to like this, but please do not accept my suggestion if you are being consistent without `()`

File: tests/functional/README-functional.md [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+This implicit change of directory simplifies the testing process and ensures that `CytoSnake` runs smoothly within the expected environment.
+
+### Output Check
+
+The output check step is for verifying if the expected outputs of your tests are generated.
+As a user, you can perform checks for multiple outputs to ensure a passing test.
+For instance, you can validate if the correct error is being raised by utilizing one of the test utility functions called `get_raised_error` (please refer to the documentation for more details).
+This helps you to thoroughly examine different aspects of the test results and validate the accuracy of `CytoSnake`'s behavior.
+
+## Ending Remarking
```
Comment (gwaybio) [inline tests/functional/README-functional.md:128]:
```suggestion
## Ending Remarks
```

File: tests/workflow/README-workflow-tests.md [hunk 30]
```diff
@@ -0,0 +1,3 @@
+# Workflow tests
+
+Docuemtnation comming soon.
```
Comment (gwaybio) [inline tests/workflow/README-workflow-tests.md:3]:
```suggestion
Documentation coming soon.
```

File: tests/unit/README-unittests.md [hunk 31]
```diff
@@ -0,0 +1,3 @@
+# Unittest documentation
+
+Comming soon.
```
Comment (gwaybio) [inline tests/unit/README-unittests.md:3]:
```suggestion
Documentation coming soon.
```

**Review by gwaybio — APPROVED** [review]
This looks great @axiomcura !

I made several in-line suggestions, mostly aiming to improve clarity and consistency. I also had some comments on small improvements to the figures. Also, the figures are not rendering in the testing rtd (but maybe this is expected until after merge)

Is the strategy to add the remaining unittest and workflow test documentation as you actually add their implementation alongside?

No need for me to re-review prior to your merging, but please don't hesitate to ask any questions.

**Review by d33bs — APPROVED** [review]
Nice work! Overall I thought things looked good and left a few comments for your consideration. Please don't hesitate to let me know if you have any questions.

File: tests/functional/README-functional.md [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+Doing so will enable other developers in the `CytoSnake` community to fully comprehend the implemented functional tests.
+
+### Understanding the main steps of CytoSnake's Functional Tests
+
+The process of creating a functional test is straightforward, involving three key steps that developers must follow to build a successful and robust test.
+These steps consist of:
+
+- **preparation**: Preparing testing folder, dataset, and selecting inputs
+- **execution**: Execute `CytoSnake` within testing older
+- **assertion checks**: Checking desired outputs
```
[the remark this answers is not in the document]
Comment (axiomcura) [reply tests/functional/README-functional.md:95]: I think anchor links are automatically generated when the website is compiled.

**Comment by axiomcura** [comment]
Applied all changes! Merging! Thank you all !
